shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/dml/impl/ShardingSelectStatementValidator.java

```diff
@@ -40,6 +41,11 @@ public void preValidate(final ShardingRule shardingRule, final SQLStatementConte
         if (isNeedMergeShardingValues(sqlStatementContext, shardingRule)) {
             needCheckDatabaseInstance = checkSubqueryShardingValues(shardingRule, sqlStatementContext, parameters, schema);
         }
+        if (!sqlStatementContext.getSqlStatement().getUnionSegments().isEmpty()) {
+            if (!shardingRule.isAllBindingTables(sqlStatementContext.getTablesContext().getTableNames()) && !shardingRule.isAllSingleTables(sqlStatementContext.getTablesContext().getTableNames())) {
+                throw new ShardingSphereException("UNION only support all binding tables or all single tables.");
```

Comment:
@tuichenchuxin `SELECT ... UNION statement can not support none binding tables.` may be better.